### DISCUSSION TITLE
Using yq4 to check for child element?

### DISCUSSION THREAD

**Discussion by simon-anz** [mikefarah/yq#734] [Q&A] [answered]
Have a structure

---
items:
- type: t1
  metadata: m1
- type: t2
  metadata: m2
- type: t3
---

Want to emit item.type only when metadata is **not** present.

When present seems to work OK

$ yq4 eval '.items[] | select(.metadata) .type'  dummy.yaml
t1
t2

But couldn't find how to invert selector.  Using length looked like it should work:

$ yq4 eval '.items[] | .metadata | length == 0'  dummy.yaml
false
false
true

but didn't??

$ yq4 eval '.items[] | select(.metadata | length == 0) .type' dummy.yaml
$

**Comment by simon-anz** [comment]
Ah, I just found "has", and when combined with |not, seems to do OK.

$ yq4 eval '.items[] | select(has("metadata") | not) .type' dummy.yaml
t3

**Comment by simon-anz** [comment, marked as answer]
How might I use"delete" to remove items 1 and 2 from the list, leaving t3 in place?

**Reply by mikefarah** [reply]
```
yq e 'del(.items[] | select(has("metadata")))'  dummy.yaml
```

**Reply by simon-anz** [reply]
Thanks! Is there a way to look deeper than one level with "has"?  My contrived example is a little simpler than my real case :)

I want to leave only items that are type=Opaque AND does not have .metadata.ownerReferences

So only the second item below should be left

---
apiVersion: v1
items:
- apiVersion: v1
  metadata:
    ownerReferences:
    - apiVersion: bitnami.com/v1alpha1
  type: Opaque
- apiVersion: v1
  metadata:
    name: aegis-ingressgateway-service-account-token-gcp69
  type: Opaque
- apiVersion: v1
  metadata:
    name: aegis-ingressgateway-service-account-token-gcp69
  type: kubernetes.io/service-account-token
- apiVersion: v1
  metadata:
    name: aegis-ingressgateway-service-account-token-gcp69
    ownerReferences:
    - apiVersion: bitnami.com/v1alpha1
  type: kubernetes.io/service-account-token
---

I can do the type filter with 
'del(.items[] | select(.type=="Opaque" | not))'

And could do a single-level "has", but... 🤯

**Reply by mikefarah** [reply]
Sure, just pipe the parent expression into 'has', e.g:

```
'.items[] | select(.metadata | has("ownerReferences"))'
```

You will need to use and/or (https://mikefarah.gitbook.io/yq/operators/boolean-operators) to combine the type selection with has

**Reply by simon-anz** [reply]
WIN!

```
$ yq4 e 'del(.items[] | select(.type=="Opaque" | not)) | del(.items[] | select(.metadata | has("ownerReferences")))' dummy2.yaml
apiVersion: v1
items:
  - apiVersion: v1
    metadata:
      name: aegis-ingressgateway-service-account-token-gcp69
    type: Opaque
```

**Reply by simon-anz** [reply]
Thanks @mikefarah !!
